### PR TITLE
Added new relation type - Multi Reference

### PR DESCRIPTION
**Pull request by konadrian**
image

### CODE REVIEW

File: modules/Settings/LayoutEditor/actions/Relation.php [hunk 1]
```diff
@@ -116,18 +116,24 @@ public function updateCustomViewOrderBy(App\Request $request): void
...
-		$type = $request->getByType('type', 'Standard');
-		$response = new Vtiger_Response();
+		$fieldName = null;
+		$source = $request->getByType('source', App\Purifier::ALNUM);
+		$target = $request->getByType('target', App\Purifier::ALNUM);
+		$label = $request->getByType('label', App\Purifier::TEXT);
+		$type = $request->getByType('type', App\Purifier::STANDARD);
 
+		if ($request->has('multi_reference_field')) {
+			$referenceFieldValues = explode('::', $request->getByType('multi_reference_field', App\Purifier::TEXT));
```
Comment (rskrzypczak) [inline modules/Settings/LayoutEditor/actions/Relation.php]: getExploded()

File: modules/Vtiger/models/Relation.php [hunk 2]
```diff
@@ -44,6 +44,9 @@ class Vtiger_Relation_Model extends \App\Base
 	/** @var int Advanced reference */
 	const RELATION_AR = 3;
 
+	//** @var int Multi reference   */
```
Comment (rskrzypczak) [inline modules/Vtiger/models/Relation.php]:
```suggestion
	/** @var int Multi reference   */
```

File: modules/Vtiger/models/Relation.php [hunk 3]
```diff
@@ -44,6 +44,9 @@ class Vtiger_Relation_Model extends \App\Base
...
+	const RELATION_MF = 4;
```
Comment (rskrzypczak) [inline modules/Vtiger/models/Relation.php]: RELATION_MF RELATION_M**R** ?

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

File: modules/Vtiger/models/Relation.php (hunk 3 above)
Comment (mariuszkrzaczkowski) [reply modules/Vtiger/models/Relation.php]: RELATION_MR

File: modules/Vtiger/models/Relation.php [hunk 4]
```diff
@@ -552,7 +555,7 @@ public function privilegeToDelete(Vtiger_Record_Model $recordModel = null, int $
 				}
 				$returnVal = !$fieldModel->isMandatory() && $fieldModel->isEditable() && !$fieldModel->isEditableReadOnly() && (!$recordModel || $recordModel->isEditable());
 			}
-			if ($this->getRelationType() === static::RELATION_AR && ($fieldModel = $this->getRelationField())) {
+			if (in_array($this->getRelationType(), [ static::RELATION_AR , static::RELATION_MF])) {
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/Relation.php]: ?

File: modules/Vtiger/uitypes/MultiReference.php [hunk 5]
```diff
@@ -222,4 +222,31 @@ public function getQueryOperators()
...
+
+	/**
+	 * Loading the list of multireference fields related with module.
+	 *
+	 * @param string $sourceModule Source module name
+	 * @param string $moduleName
+	 *
+	 * @return Vtiger_Field_Model[]
+	 */
+	public static function getFieldsRelatedWithModule(string $moduleName): array
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/uitypes/MultiReference.php]: move to modules\Settings\LayoutEditor\models\Module.php

File: public_html/layouts/basic/modules/Settings/LayoutEditor/resources/LayoutEditor.js [hunk 6]
```diff
@@ -170,42 +170,62 @@ $.Class(
...
 					App.Fields.Picklist.showSelect2ElementView(data.find('select'));
 					data.find('.relLabel').val(data.find('.target option:selected').val());
 					data.on('change', '.target', function (e) {
 						let currentTarget = $(e.currentTarget);
 						data.find('.relLabel').val(currentTarget.find('option:selected').val());
 					});
 					data.find('[name="type"]').on('change', function () {
-						if ($(this).val() === 'getAttachments') {
+						let typeOption = $(this).val();
+						if (typeOption === 'getAttachments') {
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Settings/LayoutEditor/resources/LayoutEditor.js]: switch

File: modules/Vtiger/models/Relation.php (hunk 4 above)
Comment (konadrian) [reply modules/Vtiger/models/Relation.php]: For disable button, unlink relation

File: modules/Settings/LayoutEditor/actions/Relation.php [hunk 7]
```diff
@@ -116,18 +116,24 @@ public function updateCustomViewOrderBy(App\Request $request): void
...
-		$type = $request->getByType('type', 'Standard');
-		$response = new Vtiger_Response();
+		$fieldName = null;
+		$source = $request->getByType('source', App\Purifier::ALNUM);
+		$target = $request->getByType('target', App\Purifier::ALNUM);
+		$label = $request->getByType('label', App\Purifier::TEXT);
+		$type = $request->getByType('type', App\Purifier::STANDARD);
 
+		if ($request->has('multi_reference_field')) {
+			$referenceFieldValues = $request->getExploded('multi_reference_field', '::' . App\Purifier::TEXT);
```
Comment (rskrzypczak) [inline modules/Settings/LayoutEditor/actions/Relation.php]:
```suggestion
			$referenceFieldValues = $request->getExploded('multi_reference_field', '::' . App\Purifier::ALNUM);
```

File: modules/Settings/LayoutEditor/views/Index.php [hunk 8]
```diff
@@ -109,6 +109,7 @@ public function showRelatedListLayout(App\Request $request)
 		$qualifiedModule = $request->getModule(false);
 		$viewer = $this->getViewer($request);
 		$viewer->assign('SELECTED_MODULE_NAME', $sourceModule);
+		$viewer->assign('MODULE_MULTI_REFERENCE_FIELDS', Vtiger_MultiReference_UIType::getFieldsRelatedWithModule($sourceModule));
```
Comment (rskrzypczak) [inline modules/Settings/LayoutEditor/views/Index.php]: ?

File: modules/Settings/LayoutEditor/models/Module.php [hunk 9]
```diff
@@ -1144,4 +1144,31 @@ public function getFieldInstanceByName($name)
...
+
+	/**
+	 * Loading the list of multireference fields related with module.
+	 *
+	 * @param string $sourceModule Source module name
+	 * @param string $moduleName
+	 *
+	 * @return Vtiger_Field_Model[]
+	 */
+	public static function getFieldsRelatedWithModule(string $moduleName): array
```
Comment (rskrzypczak) [inline modules/Settings/LayoutEditor/models/Module.php]: ?

File: modules/Vtiger/relations/GetMultiReference.php [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+	{
+		return Vtiger_Relation_Model::RELATION_MR;
+	}
+
+	/** {@inheritdoc} */
+	public function getQuery()
+	{
+		$relationField = $this->relationModel->getRelationField();
+		$queryGenerator = $this->relationModel->getQueryGenerator();
+		$queryGenerator->addCondition($relationField->getName(), $this->relationModel->getParentRecord()->getId(), 'c', false);
```
Comment (rskrzypczak) [inline modules/Vtiger/relations/GetMultiReference.php]: ?

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

**Pull request merged**